File: 2026/07-July/13-Monday/solution.py
from collections import deque

class Solution:
    def hasValidPath(self, grid):
        rows, cols = len(grid), len(grid[0])

        directions = {
            1: [(0, -1), (0, 1)],
            2: [(-1, 0), (1, 0)],
            3: [(0, -1), (1, 0)],
            4: [(0, 1), (1, 0)],
            5: [(0, -1), (-1, 0)],
            6: [(0, 1), (-1, 0)]
        }

        opposite = {
            (0, -1): (0, 1),
            (0, 1): (0, -1),
            (-1, 0): (1, 0),
            (1, 0): (-1, 0)
        }

        seen = [[False] * cols for _ in range(rows)]
        q = deque([(0, 0)])
        seen[0][0] = True

        while q:
            x, y = q.popleft()

            if x == rows - 1 and y == cols - 1:
                return True

            for dx, dy in directions[grid[x][y]]:
                nx, ny = x + dx, y + dy

                if 0 <= nx < rows and 0 <= ny < cols and not seen[nx][ny]:
                    if opposite[(dx, dy)] in directions[grid[nx][ny]]:
                        seen[nx][ny] = True
                        q.append((nx, ny))

        return False

File: 2026/07-July/13-Monday/solution.java
import java.util.*;

class Solution {

    private final int[][][] directions = {
            {},
            {{0, -1}, {0, 1}},
            {{-1, 0}, {1, 0}},
            {{0, -1}, {1, 0}},
            {{0, 1}, {1, 0}},
            {{0, -1}, {-1, 0}},
            {{0, 1}, {-1, 0}}
    };

    public boolean hasValidPath(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        boolean[][] visited = new boolean[rows][cols];
        Queue<int[]> queue = new LinkedList<>();

        queue.offer(new int[]{0, 0});
        visited[0][0] = true;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];

            if (x == rows - 1 && y == cols - 1)
                return true;

            for (int[] dir : directions[grid[x][y]]) {
                int nx = x + dir[0];
                int ny = y + dir[1];

                if (nx < 0 || ny < 0 || nx >= rows || ny >= cols)
                    continue;

                if (visited[nx][ny])
                    continue;

                int backX = -dir[0];
                int backY = -dir[1];

                for (int[] rev : directions[grid[nx][ny]]) {
                    if (rev[0] == backX && rev[1] == backY) {
                        visited[nx][ny] = true;
                        queue.offer(new int[]{nx, ny});
                        break;
                    }
                }
            }
        }

        return false;
    }
}
